### PR TITLE
Raise BackendStoreError when trying to insert Redis str value larger than 512MB

### LINKED ISSUE #6533: Workers retry Redis connection when task result is larger than 512MB - retry is useless as it never fits to Redis
# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [x] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.
- [x] I have included the contents of ``pip freeze`` in the issue.
- [x] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues

- None

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**: 5.0.4 (singularity)

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
software -> celery:5.0.4 (singularity) kombu:5.0.2 py:3.9.0
            billiard:3.6.3.0 py-amqp:5.0.2
platform -> system:Linux arch:64bit, ELF
            kernel version:5.9.12-arch1-1 imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:redis://:**@****:6379/10

broker_url: 'amqp://**:********@*****:5672/**'
result_backend: 'redis://:********@*****:6379/10'
deprecated_settings: None
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==5.0.2
billiard==3.6.3.0
celery @ git+https://github.com/celery/celery.git@420e3931a63538bd225ef57916deccf53cbcb57a  // == master, tried also 5.0.4
click==7.1.2
click-didyoumean==0.0.3
click-plugins==1.1.1
click-repl==0.1.6
kombu==5.0.2
prompt-toolkit==3.0.8
pytz==2020.4
redis==3.5.3
six==1.15.0
vine==5.0.0
wcwidth==0.2.5
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
#!/usr/bin/env python3

from celery import Celery

app = Celery(
    'tasks',
    broker='amqp://user:***@**:5672/**',
    backend='redis://:**@**:6379/1',
)

@app.task(ignore_result=False)
def test(*args, **kwargs):
    return 'x' * 536870911  # 512MB
```

</p>
</details>

# Expected Behavior

I'd except an exception or error. There is no point in retrying storing result to Redis when it simply never fits there. String limit is 512 MB.

I could check size of the data I'm returning in task. However, Celery adds additional metadata to my result so I do not know to what size should I limit my result.

# Actual Behavior

```
[2020-12-09 08:38:52,786: ERROR/ForkPoolWorker-8] Connection to Redis lost: Retry (0/20) now.
[2020-12-09 08:38:53,328: ERROR/ForkPoolWorker-8] Connection to Redis lost: Retry (1/20) in 1.00 second.
[2020-12-09 08:38:54,940: ERROR/ForkPoolWorker-8] Connection to Redis lost: Retry (2/20) in 1.00 second.
[2020-12-09 08:38:56,547: ERROR/ForkPoolWorker-8] Connection to Redis lost: Retry (3/20) in 1.00 second.
.... and so on
```

Redis fails with the following error when you try to set string larger than 512 MB:
```
raise ConnectionError("Error %s while writing to socket. %s." %
redis.exceptions.ConnectionError: Error 32 while writing to socket. Broken pipe.
```
Tried via another script with direct Redis connection. This says nothing about input size therefore I think value size should be checked before sending data to Redis.

### PR DESCRIPTION
**Pull request by enkelli**
## Description

Fixes #6533
TL;DR; - raise error, instead of retry reconnect to Redis when inserting value larger than max Redis string size (512MB) - retry won't help in this case.

### CODE REVIEW

**Review by auvipy — APPROVED** [review]
Looks good

**Comment by auvipy** [comment]
Pypy didn't fail in the PR but failing after merge https://github.com/celery/celery/runs/1859234610?check_suite_focus=true can you please check & consider some adjustment for the pypy3 failing test?

**Comment by enkelli** [comment]
> Pypy didn't fail in the PR but failing after merge https://github.com/celery/celery/runs/1859234610?check_suite_focus=true can you please check & consider some adjustment for the pypy3 failing test?

Hi, it does not fail locally. I see that it passed after your latest commit in `master` so may I consider it fixed?

**Comment by enkelli** [comment]
Sorry, I see the warning now, I will look into it :)

**Comment by auvipy** [comment]
yup, you are right.
